### PR TITLE
Feature/rdm 13106 search party collection field name validation

### PR DESCRIPTION
**Pull request by sem-hmcts**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RDM-13106

### Change description ###
Validation of SearchPartyTab property fields when CollectionFieldName is in use

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Review by mattnayler — CHANGES_REQUESTED** [review]

File: excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/DotNotationValidator.java [hunk 1]
```diff
@@ -21,6 +25,14 @@
     protected static final String ERROR_MESSAGE = "%sTab Invalid value '%s' is not a valid %s value. "
         + "The expression dot notation values should be valid caseTypes fields.";
 
+    public static final BiFunction<String, String, String[]> SPIT_FUNCTION =
+        (expression, separator) -> Optional.ofNullable(expression)
+            .map(x -> x.split(Pattern.quote(separator), -1))
+            .orElse(new String[0]);
+
+    public static final Function<String, String[]> DOT_SEPARATOR_SPLIT_FUNCTION =
+        expression -> SPIT_FUNCTION.apply(expression, DOT_SEPARATOR);
```
[the remark this answers is not in the document]
Comment (mattnayler) [reply excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/DotNotationValidator.java]: Resolved in e4801dcc4b856a6eabbd6655e61eb00fff987fa2

File: excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/DotNotationValidator.java [hunk 2]
```diff
@@ -58,39 +77,60 @@ private void checkDotNotationField(ParseContext parseContext,
                                        ColumnName columnName,
                                        String caseType,
                                        String expression) {
-
         try {
             // use split with -1 limit to force inclusion of empty values
-            final String[] splitDotNotationExpression = expression.split(Pattern.quote(DOT_SEPARATOR), -1);
+            final String[] splitDotNotationExpression = DOT_SEPARATOR_SPLIT_FUNCTION.apply(expression);
```
[the remark this answers is not in the document]
Comment (mattnayler) [reply excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/DotNotationValidator.java:106]: Resolved in e4801dcc4b856a6eabbd6655e61eb00fff987fa2

File: excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/DotNotationValidator.java [hunk 3]
```diff
@@ -58,39 +77,60 @@ private void checkDotNotationField(ParseContext parseContext,
...
+            // NB: start from depth = 1 to ignore top level field that is already processed
+            final List<String> segments = Arrays.asList(splitDotNotationExpression)
+                .subList(1, splitDotNotationExpression.length);
```
[the remark this answers is not in the document]
Comment (mattnayler) [reply excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/DotNotationValidator.java]: Resolved in e4801dcc4b856a6eabbd6655e61eb00fff987fa2

File: excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/DotNotationValidator.java [hunk 4]
```diff
@@ -58,39 +77,60 @@ private void checkDotNotationField(ParseContext parseContext,
...
 
             final FieldTypeEntity fieldType =
                 getTopLevelField(parseContext, sheetName, columnName, caseType, splitDotNotationExpression[0]);
 
-            Set<ComplexFieldEntity> complexFieldsBelongingToParent = fieldType.getComplexFields();
+            final Set<ComplexFieldEntity> complexFieldsBelongingToParent = fieldType.getComplexFields();
```
[the remark this answers is not in the document]
Comment (mattnayler) [reply excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/DotNotationValidator.java]: Resolved in e4801dcc4b856a6eabbd6655e61eb00fff987fa2

File: excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/DotNotationValidator.java [hunk 5]
```diff
@@ -58,39 +77,60 @@ private void checkDotNotationField(ParseContext parseContext,
...
-                }
-            }
+    public void checkDotNotationField(final Set<ComplexFieldEntity> parentComplexFields,
+                                      final SheetName sheetName,
+                                      final ColumnName columnName,
+                                      final String expression) {
+        try {
+            final String[] splitDotNotationExpression = DOT_SEPARATOR_SPLIT_FUNCTION.apply(expression);
 
+            performCheck(splitDotNotationExpression, parentComplexFields, sheetName, columnName);
```
[the remark this answers is not in the document]
Comment (mattnayler) [reply excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/DotNotationValidator.java]: Resolved in e4801dcc4b856a6eabbd6655e61eb00fff987fa2

File: excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/SearchPartyValidator.java [hunk 6]
```diff
@@ -30,7 +47,10 @@ public void validate(final List<SearchPartyEntity> searchPartyEntities,
             String caseType = searchPartyEntity.getCaseType().getReference();
 
             if (StringUtils.isNoneBlank(searchPartyEntity.getSearchPartyCollectionFieldName())) {
+                // Is this validation #validateSearchPartyCollectionFieldName() still needed?
```
[the remark this answers is not in the document]
Comment (mattnayler) [reply excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/SearchPartyValidator.java]: Resolved in e4801dcc4b856a6eabbd6655e61eb00fff987fa2

File: excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/SearchPartyValidator.java [hunk 7]
```diff
@@ -43,19 +63,86 @@ public void validate(final List<SearchPartyEntity> searchPartyEntities,
...
+
+        Set<ComplexFieldEntity> complexFields = result.getComplexFields();
+        final List<String> attributes = Arrays.asList(segments)
+            .subList(1, segments.length);
+
+        for (String segment : attributes) {
+            final ComplexFieldEntity complexFieldEntity = findComplexFieldEntity(complexFields, segment, expression);
+
+            result = complexFieldEntity.getFieldType();
+        }
```
[the remark this answers is not in the document]
Comment (mattnayler) [reply excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/SearchPartyValidator.java]: Resolved in e4801dcc4b856a6eabbd6655e61eb00fff987fa2, refactored with `DotNotationValidator`

File: excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/SearchPartyValidator.java [hunk 8]
```diff
@@ -3,18 +3,35 @@
...
+import java.util.function.Function;
+
+import static uk.gov.hmcts.ccd.definition.store.excel.validation.DotNotationValidator.DOT_SEPARATOR_SPLIT_FUNCTION;
+import static uk.gov.hmcts.ccd.definition.store.excel.validation.DotNotationValidator.SPIT_FUNCTION;
 
 @Component
 public class SearchPartyValidator {
+    private static final String COMMA_SEPARATOR = ",";
+
+    private static final String COLLECTION_FIELD_TYPE = "Collection";
```
[the remark this answers is not in the document]
Comment (mattnayler) [reply excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/SearchPartyValidator.java]: Resolved in e4801dcc4b856a6eabbd6655e61eb00fff987fa2, used `FieldTypeUtils.BASE_COLLECTION`

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/validation/SearchPartyValidatorTest.java [hunk 9]
```diff
@@ -50,6 +56,8 @@
 
     private static final String TEST_EXPRESSION_BAD = "BadTestExpression";
 
+    private static final String COLLECTION_DATA_TYPE = "Collection";
```
[the remark this answers is not in the document]
Comment (mattnayler) [reply excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/validation/SearchPartyValidatorTest.java]: Resolved in e4801dcc4b856a6eabbd6655e61eb00fff987fa2, used `FieldTypeUtils.BASE_COLLECTION`

File: aat/src/aat/resources/features/F-099 SearchParty (GlobalSearch)/F-099.feature [hunk 10]
```diff
@@ -167,3 +167,170 @@ Feature: F-099: Search Party (Global Search)
...
+  @S-099.16
+  Scenario: Import Definition file with a non-complex type referenced in the "SearchPartyCollectionFieldName" column of the SearchParty tab
+    Given a user with [an active profile in CCD]
+    When a request is prepared with appropriate values
+    And the request [contains correctly configured SearchParty tab with valid case field names in the SearchPartyName column]
+    And the request [contains correctly configured SearchParty tab with valid case field names in SearchPartyEmailAddress, SearchPartyDoB, SearchPartyAddressLine1 and SearchPartyPostCode columns]
+    And the request [contains non-complex type referenced in the SearchPartyCollectionFieldName column]
+    And it is submitted to call the [Import definition file] operation of [CCD Definition Store]
+    Then a negative response is received
+    And the response has all other details as expected
```
[the remark this answers is not in the document]
Comment (mattnayler) [reply aat/src/aat/resources/features/F-099 SearchParty (GlobalSearch)/F-099.feature:192]: Resolved in e4801dcc4b856a6eabbd6655e61eb00fff987fa2

**Review by mattnayler — APPROVED** [review]
👍 LGTM

File: excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/DotNotationValidator.java [hunk 11]
```diff
@@ -21,6 +25,14 @@
     protected static final String ERROR_MESSAGE = "%sTab Invalid value '%s' is not a valid %s value. "
         + "The expression dot notation values should be valid caseTypes fields.";
 
+    public static final BiFunction<String, String, String[]> SPIT_FUNCTION =
```
[the remark this answers is not in the document]
Comment (KleoG) [reply excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/DotNotationValidator.java]: 👍

File: excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/DotNotationValidator.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (KleoG) [reply excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/DotNotationValidator.java]: 👍

File: excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/DotNotationValidator.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (KleoG) [reply excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/DotNotationValidator.java:106]: 👍

File: excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/DotNotationValidator.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (KleoG) [reply excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/DotNotationValidator.java]: 👍

File: aat/src/aat/resources/features/F-099 SearchParty (GlobalSearch)/F-099.feature (hunk 10 above)
[the remark this answers is not in the document]
Comment (KleoG) [reply aat/src/aat/resources/features/F-099 SearchParty (GlobalSearch)/F-099.feature:192]: 👍

File: excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/DotNotationValidator.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (KleoG) [reply excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/DotNotationValidator.java]: 👍

File: excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/SearchPartyValidator.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (KleoG) [reply excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/SearchPartyValidator.java]: 👍

File: excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/SearchPartyValidator.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (KleoG) [reply excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/SearchPartyValidator.java]: 👍

File: aat/src/aat/resources/features/F-099 SearchParty (GlobalSearch)/F-099.feature [hunk 12]
```diff
@@ -167,3 +167,170 @@ Feature: F-099: Search Party (Global Search)
...
+  @S-099.15
+  Scenario: Import Definition file with a non-collection type in the "SearchPartyCollectionFieldName" column of the SearchParty tab
+    Given a user with [an active profile in CCD]
+    When a request is prepared with appropriate values
+    And the request [contains correctly configured SearchParty tab with valid case field names in the SearchPartyName column]
+    And the request [contains correctly configured SearchParty tab with valid case field names in SearchPartyEmailAddress, SearchPartyDoB, SearchPartyAddressLine1 and SearchPartyPostCode columns]
+    And the request [contains non-collection type in the SearchPartyCollectionFieldName column]
+    And it is submitted to call the [Import definition file] operation of [CCD Definition Store]
+    Then a negative response is received
+    And the response has all other details as expected
```
[the remark this answers is not in the document]
Comment (KleoG) [reply aat/src/aat/resources/features/F-099 SearchParty (GlobalSearch)/F-099.feature:180]: 👍

File: excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/validation/SearchPartyValidatorTest.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (KleoG) [reply excel-importer/src/test/java/uk/gov/hmcts/ccd/definition/store/excel/validation/SearchPartyValidatorTest.java]: 👍

File: excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/SearchPartyValidator.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (KleoG) [reply excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/SearchPartyValidator.java]: 👍

File: excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/DotNotationValidator.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (KleoG) [reply excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/DotNotationValidator.java]: 👍

File: excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/SearchPartyValidator.java [hunk 13]
```diff
@@ -43,19 +63,86 @@ public void validate(final List<SearchPartyEntity> searchPartyEntities,
...
+        FieldTypeEntity result = parseContext.getCaseFieldType(caseType, segments[0]);
+
+        Set<ComplexFieldEntity> complexFields = result.getComplexFields();
+        final List<String> attributes = Arrays.asList(segments)
+            .subList(1, segments.length);
+
+        for (String segment : attributes) {
+            final ComplexFieldEntity complexFieldEntity = findComplexFieldEntity(complexFields, segment, expression);
+
+            result = complexFieldEntity.getFieldType();
```
[the remark this answers is not in the document]
Comment (KleoG) [reply excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/validation/SearchPartyValidator.java]: 👍

**Review by KleoG — APPROVED** [review]
LGTM 👍

**Pull request merged**